lib/dynamoid_advanced_where/query_materializer.rb

```diff
@@ -23,49 +24,72 @@ def all
     def each(&blk)
       return enum_for(:each) unless blk
 
+      each_page.flat_map{|i, _| i }.each(&blk)
+    end
+
+    def each_page
       if must_scan?
-        each_via_scan(&blk)
+        each_page_via_scan
       else
-        each_via_query(&blk)
+        each_page_via_query
       end
     end
 
-    def each_via_query
+    def each_page_via_query
       query = {
-        table_name: table_name,
+        table_name: table_name
       }.merge(filter_builder.to_query_filter)
 
-      results = client.query(query)
+      Enumerator.new do |yielder|
+        start_key = nil
+        loop do
+          results = client.query(query.merge(exclusive_start_key: start_key))
+
+          start_key = results.last_evaluated_key
+
+          items = (results.items || []).map do |i|
+            klass.from_database(i.symbolize_keys)
+          end
 
-      if results.items
-        results.items.each do |item|
-          yield klass.from_database(item.symbolize_keys)
+          yielder.yield(items, results)
+
+          break if start_key.nil?
```

Comment:
Is this actually the last page behavior?

Reply:
Should we be merging this if we dont know the answer to this? Do you just need a moderate sized data set to test against ?

Reply:
probably not :D